rasa/utils/tensorflow/models.py

```diff
@@ -320,47 +155,58 @@ def train_on_batch(
-        self.prepare_for_predict()
-        self._predict_function = self._get_tf_call_model_function(
-            predict_data.as_tf_dataset, self.batch_predict, eager, "prediction"
-        )
+        self._training = None
 
-    def predict(self, predict_data: RasaModelData) -> Dict[Text, tf.Tensor]:
-        if self._predict_function is None:
-            logger.debug("There is no tensorflow prediction graph.")
-            self.build_for_predict(predict_data)
+        return self._get_metric_results()
 
-        # Prepare a single batch of the size of the input
-        batch_in = predict_data.prepare_batch()
+    def test_step(
+        self, batch_in: Union[Tuple[tf.Tensor], Tuple[np.ndarray]]
+    ) -> Dict[Text, float]:
+        """Train on batch."""
+        self._training = False
 
-        self._training = False  # needed for eager mode
-        return self._predict_function(batch_in)
+        prediction_loss = self.batch_loss(batch_in)
+        regularization_loss = tf.math.add_n(self.losses)
+        total_loss = prediction_loss + regularization_loss
+        self.total_loss.update_state(total_loss)
 
-    def save(self, model_file_name: Text, overwrite: bool = True) -> None:
-        self.save_weights(model_file_name, overwrite=overwrite, save_format="tf")
+        self._training = None
 
-    def copy_best(self, model_file_name: Text) -> None:
-        checkpoint_directory, checkpoint_file = os.path.split(self.best_model_file)
-        checkpoint_path = Path(checkpoint_directory)
+        return self._get_metric_results()
 
-        # Copy all tf2 model files from the temp location to the final destination
-        for f in checkpoint_path.glob(f"{checkpoint_file}*"):
-            shutil.move(str(f.absolute()), model_file_name + f.suffix)
+    def predict_step(
+        self, batch_in: Union[Tuple[tf.Tensor], Tuple[np.ndarray]]
+    ) -> Dict[Text, tf.Tensor]:
+        """Predict on batch."""
+        self._training = False
 
-        # Generate the tf2 checkpoint file, copy+replace to ensure consistency
-        destination_path, destination_file = os.path.split(model_file_name)
-        with open(os.path.join(checkpoint_directory, "checkpoint")) as in_file, open(
-            os.path.join(destination_path, "checkpoint"), "w"
-        ) as out_file:
-            for line in in_file:
-                out_file.write(line.replace(checkpoint_file, destination_file))
+        if not self.prepared_for_prediction:
+            # in case the model is used for prediction without loading, e.g. directly
+            # after training, we need to prepare the model for prediction once
+            self.prepare_for_predict()
+            self.prepared_for_prediction = True
 
-        # Remove the old file
-        checkpoint_path.joinpath("checkpoint").unlink()
+        return self.batch_predict(batch_in)
+
+    def _get_metric_results(self, prefix: Optional[Text] = None) -> Dict[Text, float]:
+        """Get the metrics results."""
```

Comment:
could you please update docstring with args, or remove it, since it is private method